OrbitVulkanLayer/VulkanLayerControllerTest.cpp

```diff
@@ -0,0 +1,770 @@
+                                    .pfnNextGetInstanceProcAddr = mock_get_instance_proc_addr,
+                                    .pNext = &layer_link_1};
+  VkLayerDeviceCreateInfo layer_create_info{.sType = VK_STRUCTURE_TYPE_LOADER_DEVICE_CREATE_INFO,
+                                            .function = VK_LAYER_LINK_INFO,
+                                            .u.pLayerInfo = &layer_link_2};
+  VkDeviceCreateInfo create_info{.sType = VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO,
+                                 .pNext = &layer_create_info};
+  VkDevice created_device;
+  VkPhysicalDevice physical_device = {};
+  VkResult result =
+      controller_.OnCreateDevice(physical_device, &create_info, nullptr, &created_device);
+  EXPECT_EQ(result, VK_SUCCESS);
+  EXPECT_EQ(layer_create_info.u.pLayerInfo, &layer_link_1);
+}
+
+TEST_F(VulkanLayerControllerTest, CallInDispatchTableOnGetInstanceProcAddr) {
+  const MockDispatchTable* dispatch_table = controller_.dispatch_table();
+  static constexpr PFN_vkVoidFunction kExpectedFunction = +[]() {};
+  PFN_vkGetInstanceProcAddr mock_get_instance_proc_addr =
+      +[](VkInstance /*instance*/, const char* /*name*/) { return kExpectedFunction; };
+  EXPECT_CALL(*dispatch_table, GetInstanceProcAddr)
+      .Times(1)
+      .WillOnce(Return(mock_get_instance_proc_addr));
+  VkInstance instance = {};
+  PFN_vkVoidFunction result = controller_.OnGetInstanceProcAddr(instance, "some function");
+  EXPECT_EQ(result, kExpectedFunction);
+}
+
+TEST_F(VulkanLayerControllerTest, WillClearUpOnDestroyInstance) {
+  PFN_vkDestroyInstance mock_destroy_instance =
+      +[](VkInstance /*instance*/, const VkAllocationCallbacks* /*allocator*/) {};
+  const MockDispatchTable* dispatch_table = controller_.dispatch_table();
+  EXPECT_CALL(*dispatch_table, DestroyInstance).Times(1).WillOnce(Return(mock_destroy_instance));
+  EXPECT_CALL(*dispatch_table, RemoveInstanceDispatchTable).Times(1);
+
+  VkInstance instance = {};
+
+  controller_.OnDestroyInstance(instance, nullptr);
+}
+
+TEST_F(VulkanLayerControllerTest, WillClearUpOnDestroyDevice) {
+  PFN_vkDestroyDevice mock_destroy_device =
+      +[](VkDevice /*device*/, const VkAllocationCallbacks* /*allocator*/) {};
+  const MockDispatchTable* dispatch_table = controller_.dispatch_table();
+  EXPECT_CALL(*dispatch_table, DestroyDevice).Times(1).WillOnce(Return(mock_destroy_device));
+  EXPECT_CALL(*dispatch_table, RemoveDeviceDispatchTable).Times(1);
+  const MockDeviceManager* device_manager = controller_.device_manager();
+  EXPECT_CALL(*device_manager, UntrackLogicalDevice).Times(1);
+
+  VkDevice device = {};
+
+  controller_.OnDestroyDevice(device, nullptr);
+}
+
+// ----------------------------------------------------------------------------
+// Core layer logic
+// ----------------------------------------------------------------------------
+
+TEST_F(VulkanLayerControllerTest, CanDelegateOnResetCommandPool) {
```

Comment:
Is delegate the right word here? Maybe "forward", as `SubmissionTracker::ResetCommandPool` is not the only work done in the system on `OnResetCommandPool` (or is it)? Also why "can"? And maybe specify the receiving end? In other words, if I get the test correctly this would be something like `ForwardsOnResetCommandPoolToSubmissionTracker`? Similarly for other tests below.

Reply:
Yeah that is better. Thanks for the suggestion.